great_expectations/core/batch_definition.py

```diff
@@ -76,14 +82,70 @@ def get_batch(self, batch_parameters: Optional[BatchParameters] = None) -> Batch
         batch_request = self.build_batch_request(batch_parameters=batch_parameters)
         return self.data_asset.get_batch(batch_request)
 
-    def is_added(self) -> BatchDefinitionAddedDiagnostics:
-        return BatchDefinitionAddedDiagnostics(
+    def is_fresh(self) -> BatchDefinitionFreshnessDiagnostics:
+        diagnostics = self._is_added()
+        if not diagnostics.success:
+            return diagnostics
+        return self._is_fresh()
+
+    def _is_added(self) -> BatchDefinitionFreshnessDiagnostics:
+        return BatchDefinitionFreshnessDiagnostics(
             errors=[] if self.id else [BatchDefinitionNotAddedError(name=self.name)]
         )
 
+    def _is_fresh(self) -> BatchDefinitionFreshnessDiagnostics:
+        datasource_dict = project_manager.get_datasources()
+
+        datasource: Datasource | None = None
+        try:
+            datasource = datasource_dict[self.data_asset.datasource.name]
+        except KeyError:
+            pass
```

Comment:
This logic needs tests - does it make sense to capture all these errors in aggregate or should we panic and alert the user upon hitting the first one?